Add riscv64 baremetal runtime

ElEHsiang opened the pull request:
This patch add a build rule for riscv64 baremetal runtime.

The riscv64 baremetal toolchain can get from https://github.com/riscv-collab/riscv-gnu-toolchain.
riscv64 baremetal toolchain build command:
```
git clone https://github.com/riscv-collab/riscv-gnu-toolchain.git
cd riscv-gnu-toolchain
mkdir -p build/install && cd build
../configure --enable-llvm --prefix=`realpath install`
make -j8
```

Build command
```
cmake -G Ninja \
    -B ../iree-baremetal-rv64 \
    -DCMAKE_TOOLCHAIN_FILE="./build_tools/cmake/riscv.toolchain.cmake" \
    -DRISCV_TOOLCHAIN_ROOT=<riscv64 baremetal toolchain root> \
    -DRISCV_CPU="generic-riscv_64" \
    -DRISCV_TOOLCHAIN_PREFIX="riscv64-unknown-elf-" \
    -DIREE_HOST_BIN_DIR=<host tool bin path> \
    -DRISCV_COMPILER_FLAGS="-O2" \
    -DIREE_BUILD_COMPILER=OFF \
    -DIREE_ENABLE_CPUINFO=OFF \
    -DIREE_BUILD_TESTS=OFF \
    -DIREE_ENABLE_ASSERTIONS=ON \
    -DIREE_BUILD_SAMPLES=ON \
    .
    
cmake --build ../iree-baremetal-rv64
```

Test Command
```
qemu-riscv64 -cpu rv64 -B 0x100000  ../iree-baremetal-rv64-gnu/samples/simple_embedding/simple_embedding_embedded_sync
```

ElEHsiang committed:
Add riscv64 baretemal runtime

ScottTodd commented on build_tools/cmake/riscv.toolchain.cmake:
```diff
@@ -55,6 +56,30 @@ if(RISCV_CPU STREQUAL "linux-riscv_64")
     "--iree-llvmcpu-target-cpu-features=+m,+a,+f,+d,+c,+zvl512b,+v"
     "--riscv-v-fixed-length-vector-lmul-max=8"
     CACHE INTERNAL "Default llvm codegen flags for testing purposes")
+elseif(RISCV_CPU STREQUAL "generic-riscv_64")
```
Looks like several flags are duplicated across linux/generic and 32/64 here. Maybe we'd want to refactor?

```cmake
if(RISCV_CPU MATCHES "generic")
  # disable OS features
elseif(RISCV_CPU MATCHES "linux")
  # enable OS features
endif()

if(RISCV_CPU MATCHES "riscv_32")
  # set 32-bit flags
elseif(RISCV_CPU MATCHES "riscv_64")
  # set 64-bit flags
endif()
```

ElEHsiang replied:
Ok, I'll refactor it.

ElEHsiang replied:
I can pull duplicate code out of the this codition. But some code can't pull out. For example:
`linux-riscv_32` has `RISCV32_TEST_DEFAULT_LLVM_FLAGS`.
`generic-riscv64` has additional `RISCV_COMPILER_FLAGS`

So we still need the `if(RISCV_CPU STREQUAL "generic-riscv_64")` condition.

ScottTodd replied:
Good points. I think it could still be useful to restructure a bit, as that would make it more clear which options are shared based on a general property and which are unique to a specific configuration.

hcindyl commented on build_tools/cmake/riscv.toolchain.cmake:
```diff
@@ -55,6 +56,30 @@ if(RISCV_CPU STREQUAL "linux-riscv_64")
+  # Specify ISP spec for march=rv64gc. This is to resolve the mismatch between
+  # llvm and binutil ISA version.
+  set(RISCV_COMPILER_FLAGS "${RISCV_COMPILER_FLAGS} \
+      -march=rv64gc -mabi=lp64d -DIREE_PLATFORM_GENERIC=1 -DIREE_SYNCHRONIZATION_DISABLE_UNSAFE=1 \
```
Please spell out the ISA version, or there will be version mismatch spamming during linking time.

hcindyl replied:
ahh... the ISA version is aligned by https://github.com/llvm/llvm-project/commit/dc90af501f00bb0bbbfde2d90360f074922e3e81. However, the toolchains that we provided in IREE project (and embedded in the docker) were built before this commit. It would still be better to specify the version to avoid confusion.

ElEHsiang replied:
OK, I'll add the ISA version.

hcindyl commented on build_tools/cmake/riscv.toolchain.cmake:
```diff
@@ -23,14 +23,15 @@ set(RISCV_HOST_TAG linux)
-set(CMAKE_STRIP "${RISCV_TOOLCHAIN_ROOT}/bin/llvm-strip")
+set(CMAKE_C_COMPILER "${RISCV_TOOLCHAIN_ROOT}/bin/${RISCV_TOOLCHAIN_PREFIX}clang")
+set(CMAKE_CXX_COMPILER "${RISCV_TOOLCHAIN_ROOT}/bin/${RISCV_TOOLCHAIN_PREFIX}clang++")
+set(CMAKE_AR "${RISCV_TOOLCHAIN_ROOT}/bin/${RISCV_TOOLCHAIN_PREFIX}ar")
```
```suggestion
set(CMAKE_AR "${RISCV_TOOLCHAIN_ROOT}/bin/${RISCV_TOOLCHAIN_PREFIX}llvm-ar")
```
and same as the tools below

ElEHsiang replied:
If we use the toolsuite built from [riscv-collab/riscv-gnu-toolchain](https://github.com/riscv-collab/riscv-gnu-toolchain) , there is no llvm binutils. Should I add a rule to build llvm binutils in riscv-collab/riscv-gnu-toolchain? Or can we update the toolsuite provided by IREE?

hcindyl replied:
IREE's RISCV toolchains were built with  the clang + gcc built flow as discribed in [webdoc](https://openxla.github.io/iree/building-from-source/riscv/#install-risc-v-cross-compile-toolchain-and-emulator), thus they have both gcc and llvm binutils. The `--enable-llvm` flag you pointed in this PR was added last month, while those toolchains were built sometime last year.

Also, IREE's RISCV Linux toolchain has the [multilib](https://github.com/riscv-collab/riscv-gnu-toolchain#build-with-customized-multi-lib-configure) enabled to support both riscv64 and riscv32 builds with the same toolchain.

We could update IREE's RISCV existing toolchains, but unfortunately I don't see a strong reason to do so. If the goal is to add a support to build for riscv64 baremental, the least amount of change seems to be having the proper prefix set for the existing toolchains and the new one. At the same time, @dcaballe can chime in on if we want to formally support/maintain RISCV64 baremetal config, that is, adding the toolchain to the IREE artifact bucket, setting up CI unit test, etc. If not, maybe adding a section in [webdoc source](https://github.com/openxla/iree/blob/main/docs/website/docs/building-from-source/riscv.md) explaining how rv64-baremetal can be supported would be sufficient.

ElEHsiang replied:
Oh...the webdoc describe how to build both gcc and llvm binutils. I'll change back to use llvm binutils. Sorry for not notice this.

hcindyl commented on build_tools/cmake/riscv.toolchain.cmake:
```diff
@@ -23,14 +23,15 @@ set(RISCV_HOST_TAG linux)
 set(RISCV_TOOL_PATH "$ENV{HOME}/riscv" CACHE PATH "RISC-V tool path")
 
 set(RISCV_TOOLCHAIN_ROOT "${RISCV_TOOL_PATH}/toolchain/clang/${RISCV_HOST_TAG}/RISCV" CACHE PATH "RISC-V compiler path")
+set(RISCV_TOOLCHAIN_PREFIX "" CACHE STRING "RISC-V compiler prefix")
```
Please set a default, and also update https://github.com/openxla/iree/blob/main/build_tools/cmake/build_riscv.sh with the proper prefix

ElEHsiang replied:
Ok, I'll set a default value and update build_riscv.sh

ElEHsiang committed:
Refactor riscv.toolchain.cmake

1. Use riscv64-unknown-linux-gnu- as default RISCV_TOOLCHAIN_PREFIX
2. Revert "use gnu binutils" change
3. Refactor the setting for different target

ElEHsiang committed:
Add corresponding RISCV_TOOLCHAIN_PREFIX in build_riscv.sh

hcindyl commented on build_tools/cmake/riscv.toolchain.cmake:
```diff
@@ -23,14 +23,15 @@ set(RISCV_HOST_TAG linux)
-set(CMAKE_AR "${RISCV_TOOLCHAIN_ROOT}/bin/llvm-ar")
-set(CMAKE_RANLIB "${RISCV_TOOLCHAIN_ROOT}/bin/llvm-ranlib")
-set(CMAKE_STRIP "${RISCV_TOOLCHAIN_ROOT}/bin/llvm-strip")
+set(CMAKE_C_COMPILER "${RISCV_TOOLCHAIN_ROOT}/bin/${RISCV_TOOLCHAIN_PREFIX}clang")
```
This is still not correct. The existing toolchains do not have the prefix for all the binaries.

You can test out the current toolchain/build  behavior with the docker command found in https://github.com/openxla/iree/blob/ffeaa264397dbe32127d83efce1fcf5a5930e6d4/.github/workflows/ci.yml#L829-L846

ElEHsiang replied:
I see there is no prefix for the clang* and binutil in docker. I'll remove these change. Thanks for the information.

ElEHsiang committed:
Remove wrong toolchain prefix

hcindyl reviewed (approved):
I am okay with the change; please see if @ScottTodd  or @GMNGeoffrey has other feedback. Thanks.
